File: services/api/internal/pipeline/worker.go
package pipeline

import (
	"context"
	"encoding/json"
	"log/slog"
	"time"

	"github.com/piratepod/api/internal/db"
)

type Worker struct {
	repo   *db.Repo
	client *Client
	log    *slog.Logger
	wake   chan struct{}
}

func NewWorker(repo *db.Repo, client *Client, log *slog.Logger) *Worker {
	return &Worker{
		repo:   repo,
		client: client,
		log:    log,
		wake:   make(chan struct{}, 1),
	}
}

func (w *Worker) Wake() {
	select {
	case w.wake <- struct{}{}:
	default:
	}
}

func (w *Worker) Start(ctx context.Context) {
	go w.loop(ctx)
}

func (w *Worker) loop(ctx context.Context) {
	if err := w.repo.ResetRunningToQueued(ctx); err != nil {
		w.log.Error("reset running jobs", slog.Any("err", err))
	}
	ticker := time.NewTicker(time.Second)
	defer ticker.Stop()
	for {
		if err := w.drain(ctx); err != nil {
			w.log.Error("job drain", slog.Any("err", err))
		}
		select {
		case <-ctx.Done():
			return
		case <-w.wake:
		case <-ticker.C:
		}
	}
}

func (w *Worker) drain(ctx context.Context) error {
	for {
		job, err := w.repo.ClaimNextQueued(ctx)
		if err != nil || job == nil {
			return err
		}
		w.process(ctx, *job)
	}
}

func (w *Worker) process(ctx context.Context, job db.Job) {
	log := w.log.With(slog.String("job_id", job.ID))
	var req GenerateRequest
	if err := json.Unmarshal([]byte(job.URLsJSON), &req); err != nil {
		_ = w.repo.SetFailed(ctx, job.ID, db.StageQueued, "invalid stored job request")
		log.Error("decode job request", slog.Any("err", err))
		return
	}
	setStage := func(stage, message string) error {
		log.Info("job stage", slog.String("stage", stage), slog.String("message", message))
		return w.repo.SetRunningStage(ctx, job.ID, stage, message)
	}
	result, err := w.client.Generate(ctx, req, setStage)
	if err != nil {
		stage := db.StageQueued
		latest, getErr := w.repo.GetJob(ctx, job.ID, job.OwnerID)
		if getErr == nil {
			stage = latest.Stage
		}
		if setErr := w.repo.SetFailed(ctx, job.ID, stage, err.Error()); setErr != nil {
			log.Error("set failed", slog.Any("err", setErr))
		}
		log.Error("job failed", slog.Any("err", err))
		return
	}
	payload, err := json.Marshal(result)
	if err != nil {
		_ = w.repo.SetFailed(ctx, job.ID, db.StageDone, "failed to encode generation result")
		log.Error("encode result", slog.Any("err", err))
		return
	}
	if err := w.repo.SetSucceeded(ctx, job.ID, string(payload)); err != nil {
		log.Error("set succeeded", slog.Any("err", err))
		return
	}
	log.Info("job succeeded", slog.String("title", result.Title), slog.String("feed_url", result.FeedURL))
}

File: workers/orchestrate/src/orchestrate/worker.py
import asyncio
import json
from collections.abc import Awaitable, Callable

from fastapi import HTTPException
from pydantic import ValidationError

from piratepod_core.logging import get_logger

from . import db
from .db import Job, JobRepo
from .schemas import GenerateRequest, GenerateResponse

Pipeline = Callable[
    [GenerateRequest, Callable[[str, str], Awaitable[None]]], Awaitable[GenerateResponse]
]

log = get_logger(__name__)


class JobWorker:
    def __init__(self, repo: JobRepo, pipeline: Pipeline, poll_interval: float = 1.0):
        self.repo = repo
        self.pipeline = pipeline
        self.poll_interval = poll_interval
        self._wake = asyncio.Event()
        self._task: asyncio.Task[None] | None = None

    def start(self) -> None:
        if self._task is None:
            self._task = asyncio.create_task(self._loop())

    async def stop(self) -> None:
        if self._task is None:
            return
        self._task.cancel()
        try:
            await self._task
        except asyncio.CancelledError:
            pass
        self._task = None

    def wake(self) -> None:
        self._wake.set()

    async def _loop(self) -> None:
        try:
            await asyncio.to_thread(self.repo.reset_running_to_queued)
        except Exception as exc:
            log.error("orchestrate.reset_running.failed", err=str(exc))
        while True:
            try:
                await self._drain()
            except Exception as exc:
                log.error("orchestrate.drain.failed", err=str(exc))
            try:
                await asyncio.wait_for(self._wake.wait(), timeout=self.poll_interval)
                self._wake.clear()
            except TimeoutError:
                pass

    async def _drain(self) -> None:
        while True:
            job = await asyncio.to_thread(self.repo.claim_next_queued)
            if job is None:
                return
            await self._process(job)

    async def _process(self, job: Job) -> None:
        log.info("orchestrate.job.start", job_id=job.id)
        try:
            req = GenerateRequest.model_validate_json(job.urls_json)
        except ValidationError:
            await asyncio.to_thread(
                self.repo.set_failed,
                job.id,
                db.STAGE_QUEUED,
                "invalid stored job request",
            )
            return

        async def set_stage(stage: str, message: str) -> None:
            log.info("orchestrate.job.stage", job_id=job.id, stage=stage)
            await asyncio.to_thread(self.repo.set_running_stage, job.id, stage, message)

        try:
            result = await self.pipeline(req, set_stage)
        except Exception as exc:
            latest = await _latest_job(self.repo, job)
            await asyncio.to_thread(
                self.repo.set_failed,
                job.id,
                latest.stage,
                _error_message(exc),
            )
            log.error("orchestrate.job.failed", job_id=job.id, err=str(exc))
            return

        payload = result.model_dump_json()
        await asyncio.to_thread(self.repo.set_succeeded, job.id, payload)
        log.info("orchestrate.job.succeeded", job_id=job.id, title=result.title)


async def _latest_job(repo: JobRepo, fallback: Job) -> Job:
    try:
        return await asyncio.to_thread(repo.get_job, fallback.id, fallback.owner_id)
    except KeyError:
        return fallback


def _error_message(exc: Exception) -> str:
    if isinstance(exc, HTTPException):
        detail = exc.detail
        if isinstance(detail, str):
            return detail
        return json.dumps(detail)
    return str(exc) or exc.__class__.__name__
